client/src/main/java/io/pravega/client/tables/TableSegment.java

```diff
@@ -0,0 +1,130 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.tables;
+
+import io.pravega.common.util.AsyncIterator;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import lombok.Data;
+
+/**
+ * Defines all operations that are supported on a Table Segment.
+ *
+ * @param <KeyT>   Table Key Type.
+ * @param <ValueT> Table Value Type.
+ */
+public interface TableSegment<KeyT, ValueT> extends AutoCloseable {
+    /**
+     * Inserts or updates an existing Table Entry into this Table Segment.
+     * @param entry The Entry to insert or update. If {@link VersionedEntry#getVersion()} is null, this will perform a
+     *              Blind Update, otherwise it will perform a Conditional Update based on the information provided.
+     * @return A CompletableFuture that, when completed, will contain the {@link KeyVersion} associated with the newly
+     * inserted or updated entry.
+     */
+    CompletableFuture<KeyVersion> put(VersionedEntry<KeyT, ValueT> entry);
```

Comment:
Based on our discussions earlier, we need a clear distinction between `create` and `update` apis on tablesegments. 
A create should inform the caller if the key already "exists".

Reply:
Between that and conditional updates, it is my opinion that it will create too many redundant APIs.

I have thus tweaked the existing API to handle this in `put`. In a nutshell:
- Blind Updates/Removals will succeed regardless of whether the Key exists or not, and regardless of what their version is.
- Conditional Updates/Removals will succeed only if the provided Key Version matches what is in the Table Segment itself.
- To make life easier, I am providing a special Version (`KeyVersion.NOT_EXISTS`) which, if provided to `put`, will create a conditional update/removal which is conditioned on the key not existing.
    - This doesn't make much sense for `remove`, but I could imagine some interesting cases where it might be useful, so I'm keeping the generalization.